MudExitPrompt: Add component

91378246 opened the pull request:
This PR adds a small component which can be used to require a confirmation before navigating away, closing or reloading.
![navconf](https://github.com/user-attachments/assets/1b04895a-9242-4679-bb92-0f4763c5a22e)

91378246 committed:
Add component

91378246 committed:
Pipeline

91378246 committed:
Pipeline

91378246 commented:
@ScarletKuro The problem is that there is no way to catch and delay beforeunload as far as I know; this is handled by the browser directly. I could in theory show a custom popup on navigating away but I deliberately didn't do that as it would then show the browser dialog on close/reload and the MudDialog on navigation. This would probably confuse devs as their dialog is shown on some actions but not on all.

Translations: Hardcoded, as I wanted both dialogs to look as similar as possible (it's just the translation of the browser dialog text). But I could also just add a string field with a default english text if that's prefered.

henon commented:
I really like the usefulness of this functionality (I might need this in my SW) but I don't like the name. Here are some suggestions:
- MudNavigationGuard
- MudNavigationPrompt
- MudNavigationAlert
- MudExitPrompt

91378246 committed:
Rename, add test, add localizer

91378246 commented:
@ScarletKuro 
- Can you take a look MudBlazor.MudExitPrompt.EnableAsync? I'm not rly sure if I use the localizer correctly
- I added a test but I wasn't able to trigger the location changing handler (`NavigationManager.RegisterLocationChangingHandler(OnLocationChanging)`) => do you know why it doesn't fire?

ScarletKuro committed:
Merge branch 'dev' into feature/confirm-navigation

ScarletKuro committed:
Fix tests

ScarletKuro committed:
dotnet format

ScarletKuro commented:
Fixed the tests. Not sure why this was necessary:
```csharp
Context.Services.AddSingleton<NavigationManager>(s => s.GetRequiredService<BunitNavigationManager>());
```
This shouldn’t have been required, since `new BunitContext()` calls `AddDefaultBunitContextServices`, which already registers this exact service. So why it wasn’t available is unclear. Some JavaScript setup wasn’t done correctly as well.

Btw:
```csharp
protected virtual async ValueTask DisposeAsyncCore()
{
    if (!Disabled && _navigatedAway)
    {
        await DisableAsync();
    }
}
```

Shouldn’t the component always be disposed when `Disabled == false`?
If `handleBeforeNavigation` returns `false` and the component is disposed, you won’t unsubscribe from `beforeunload`. Is that intentional?

91378246 commented:
Yeah, that is intentional; This would prevent the js from reacting to tab closed (that's also the reason the `_navigatedAway` flag exists)

ScarletKuro commented:
Ah okay, thanks.
Just to be transparent, in January we’ll be creating official extensions (see https://github.com/MudBlazor/MudBlazor/discussions/12323#discussioncomment-15335478) and this will probably move there. Nothing needs to be done on your side, I’ll adjust the namespaces and code location myself.

It will mostly likely live here in this mono repo, and to make things easier, docs will reference it directly.

danielchalmers commented:
Update: We think this would be a better candidate for a service instead of a component - would you be interested in reworking it into one? No problem at all if you don't want to commit extra work, sorry for the delay!

91378246 commented:
Maybe in a new section under Features called "Functional components" or something like that where we can also describe other such components like `MudThemeProvider`, `MudPopoverProvider`, etc.?

91378246 committed:
Merge branch 'dev' into feature/confirm-navigation

# Conflicts:
#	src/MudBlazor.UnitTests.Shared/Mocks/MockNavigationManager.cs
#	src/MudBlazor/Resources/LanguageResource.resx

91378246 commented:
@danielchalmers (Don't know if you are the correct ping for that) I believe there is a pipeline issue:

```
Run dotnet build src/MudBlazor.UnitTests.Docs -c Release --no-restore
  MudBlazor.Analyzers -> /home/runner/work/MudBlazor/MudBlazor/src/MudBlazor.Analyzers/bin/Release/netstandard2.0/MudBlazor.Analyzers.dll
  MudBlazor.Analyzers.CodeFixes -> /home/runner/work/MudBlazor/MudBlazor/src/MudBlazor.Analyzers.CodeFixes/bin/Release/netstandard2.0/MudBlazor.Analyzers.CodeFixes.dll
  Building MudBlazor web assets (inner build, TFM: net10.0)
  MudBlazor.Examples.Data -> /home/runner/work/MudBlazor/MudBlazor/src/MudBlazor.Examples.Data/bin/Release/net10.0/MudBlazor.Examples.Data.dll
  Error: Response status code does not indicate success: 500 (Internal Server 
  Error).
/home/runner/work/MudBlazor/MudBlazor/src/MudBlazor/MudBlazor.csproj(142,5): error MSB3073: The command "dotnet tool exec BunDotNet.Cli@1.2.0 --yes -- wrapper --version 1.3.6 --path ../../tools --silent -- run build.mjs" exited with code 255. [TargetFramework=net10.0]
```

Builds locally without any issues.

91378246 committed:
Add functional docs category and add more customization options

91378246 committed:
Pipeline

danielchalmers committed:
Merge branch 'dev' into feature/confirm-navigation

danielchalmers committed:
documentation

danielchalmers committed:
Fix test class name

danielchalmers committed:
move properties with the rest

danielchalmers committed:
refactor OnLocationChanging

danielchalmers committed:
Rename NativeOnly to UseNativePrompt

might be more intuitive

danielchalmers committed:
test comments

danielchalmers committed:
update xml docs

danielchalmers committed:
Change default exit button text to match browser

danielchalmers committed:
update docs page and example

danielchalmers commented on src/MudBlazor/Components/ExitPrompt/MudExitPrompt.razor.cs:
```diff
@@ -0,0 +1,183 @@
+
+    private async Task<bool> IsDialogNavigationAllowedAsync()
+    {
+        return await DialogService.ShowMessageBoxAsync(
+            TitleToDisplay,
+            TextToDisplay,
+            Localizer[LanguageResource.MudExitPrompt_Exit],
+            Localizer[LanguageResource.MudExitPrompt_Cancel]
+        ) == true;
```
ShowMessageBoxAsync returns a nullable bool

danielchalmers committed:
always disable on disposal

DisposeAsyncCore only calls DisableAsync when _navigatedAway is true. If the component is removed/disposed without a successful navigation event (e.g., conditionally rendered off, layout change, error boundary teardown), the beforeunload handler can remain registered and keep prompting unexpectedly. Consider always invoking DisableAsync() during disposal (it is effectively idempotent thanks to the JS-side guard/error handling).

danielchalmers commented on src/MudBlazor.UnitTests/Components/ExitPromptTests.cs:
```diff
@@ -0,0 +1,51 @@
+        // Disposal should always unregister one final time.
+        await Context.DisposeComponentsAsync();
+        jsRuntimeMock.Verify(x => x.InvokeAsync<IJSVoidResult>("mudExitPrompt.disable", It.IsAny<object[]>()), Times.Exactly(2));
+    }
```
already partly covered by DisposeWithoutNavigationDisablesPrompt

danielchalmers commented on src/MudBlazor/Components/ExitPrompt/MudExitPrompt.razor.cs:
```diff
@@ -0,0 +1,183 @@
+
+    private async Task<bool> IsNativeNavigationAllowedAsync()
+    {
+        return await JsRuntime.InvokeAsync<bool>("mudExitPrompt.handleBeforeNavigation");
```
not needed

danielchalmers committed:
Refactor JS interop from single shared state to per-instance state keyed by component ID, so concurrent/transient instances do not overwrite each other or disable each other

The JS interop is implemented as a single global singleton (window.mudExitPrompt) with a single isEnabled flag and text value. If multiple MudExitPrompt instances exist (even transiently during layout/route transitions), they can overwrite each other’s text and one instance disabling will remove the beforeunload listener for all.

danielchalmers committed:
Rewrite ExitPromptTests from scratch with a full, helper-backed coverage suite

danielchalmers committed:
Strict JS throws + C# registration tracking reintroduced

danielchalmers committed:
Added explicit tests for unregistered/disabled scenarios

danielchalmers committed:
more tests

91378246 commented:
@danielchalmers I don't quite understand why you added map in the js; Is there any use case where you would need more than one MudExitPrompt?
Also as a personal note: I think it's a bit much to move single line commands into their own functions, that makes analysing the flow very difficult, especially while using multiple different flags (at least for me).

ScarletKuro commented:
> but can the original version (no map) handle navigating from one page with an exit prompt to a different page with one as well, without any conflicts that might be caused by disposal/initialization ordering?

I’ve never seen a conflict here during page navigation. I expect Blazor to have a very strict lifecycle; otherwise, per-page interactivity with providers on each page would be hella broken

> And how should we handle the case where someone adds multiple exit prompts on the same page?

This makes no sense; I would throw an exception in this case. But detecting duplicates might add another layer of complication.

danielchalmers commented:
If we're not sure how it would work we can just leave it with the extra safety, it doesn't complicate it much anyway

danielchalmers committed:
inline into IsNavigationAllowedAsync

danielchalmers committed:
delete _isPromptEnabled and instead throw an error if it wasn't able to enable the prompt
